### DISCUSSION TITLE
Multi user support in Trame

### DISCUSSION THREAD

**Discussion by mirabdullahyaser** [Kitware/trame#32] [Q&A] [answered]
Hello,

Is it possible that multiple users can use the same trame server and each user can have their independent sessions on their respective 
machines? Currently, change done by one user also changes it for others. How can we achieve this if it is possible?

Thanks!

**Comment by jourdain** [comment, marked as answer]
The simple solution is to add `--launcher` to your command line.

Assuming it works with your setup, what will happen is that trame will automatically generate a configuration for our launcher for your application and start that launcher in a way that any browser connecting to your endpoint will get its own process with your trame application on a different port. Everything happen in the background so you don't really know that other ports are getting used for the websocket connection. 

But in a cloud deployment, you most likely want every connection to go over http(80) or https(443) regardless on how many users connects. And for that specific configuration, the best way to deploy your trame application is to use a docker image similar to the one we have created for [ParaViewWeb](https://github.com/Kitware/paraviewweb/tree/master/tools/docker/paraviewweb). 

Hopefully within a week or two we will have dedicated docker images for trame to streamline such deployment along with an update documentation about that process.

To better understand how the docker deployment works, you can read the following set of guides:
- [What is happening inside docker](http://kitware.github.io/paraviewweb/docs/multi_user_setup.html)
- [Exposing a docker setup behind https](http://kitware.github.io/paraviewweb/docs/docker.html)
- [The layer that matters](http://kitware.github.io/paraviewweb/docs/container_overview.html#Layer-Two-ParaViewWeb)

**Reply by mirabdullahyaser** [reply]
I have added the --launcher to my command line but it did not work and it is getting the wrong host and port as shown in the image below. I am using the trame version 1.13.2

![Screenshot (233)](https://user-images.githubusercontent.com/42169766/151121185-88a53667-4ed1-437b-89dd-34f3710e8465.png)

**Reply by rodrigomologni** [reply]
Hi @jourdain!

Multi-user support is not working here either. I've added `--launcher` on command line but the launcher is not starting.

```
> python .\trame-tutorial\04_application\solution.py --server --launcher

App running at:
 - Local:   http://localhost:8080/
 - Network: http://172.23.176.1:8080/

Note that for multi-users you need to use and configure a launcher.
```

![record](https://user-images.githubusercontent.com/1608652/181820077-573ea226-1819-4c45-8c40-95c43a71297e.gif)

**Reply by jourdain** [reply]
The `--launcher` arg has been removed in trame v2 as it was just an helper for building a launcher config.
Now with the cookiecutter, you get a docker image to do exactly that.

**Reply by jourdain** [reply]
Either way, the `--launcher` was never meant for production usage while the docker image is.

**Reply by lixiaoqin0218** [reply]
How can I use 'cookiecutter' for multi-user deployment? What are the specific examples?

**Reply by jourdain** [reply]
The cookiecutter provides a bundle directory with a docker subsection. Bundling within docker is what makes your app multi-user.

You can find a simpler example of it [here](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile).

**Comment by jourdain** [comment]
you are missing `--` in front of your `port` arg.

**Reply by mirabdullahyaser** [reply]
Oh my bad but I have tried it without both --port and --host arguments and --launcher as an only argument it did not work and I get the same 
======== Running on http://0.0.0.0:8080 ========
and when I gave it correctly --port and --host arguments along with --launcher it did not work in that case as well and I get
======== Running on http://0.0.0.0:1234 ========

**Reply by jourdain** [reply]
Can you explain what you mean by not work? From what I see the launcher start and wait for client connection. So part of it seems to work, now I'm not sure which part does not work.

**Reply by mirabdullahyaser** [reply]
Why it is running on 0.0.0.0 even if I specify host IP with --host it runs on 0.0.0.0 and I cannot access the trame server on the mentioned address of http://0.0.0.0:8080 When I try to access it with my systems IP I get 'session did not start before timeout expired. Check session logs'

**Reply by jourdain** [reply]
0.0.0.0 is to not limit to only one network endpoint on your host. You can still use your IP in the url of your browser.
So now, the next step for you would be to figure out why the process for your user did not start properly by looking at the log files.

Those log files will be in `/tmp` also if you provide the debug flag, the launcher configuration used will be printed to the console.

**Reply by mirabdullahyaser** [reply]
I have checked the logs and found the issue that I am running the trame application inside python venv but when I use --launcher flag and checked logs I found that now it is using my system's python and not the python venv. When I run the trame application with my system's python I get the error (same as on log file) AttributeError: 'App' object has no attribute 'server_port' though I am using same trame version both in python venv and my system's python.

**Reply by jourdain** [reply]
`pip install -U pywebvue`
